### DISCUSSION TITLE
numpy/scipy adjacency matrix to Graph

### DISCUSSION THREAD

**Discussion by Jessime** [networkit/networkit#700] [Q&A]
I have a large ~60K rows/cols adjacency matrix (either as a `np.array` or a `scipy.sparse.csr_matrix`, depending on the density of edges) that I'd like to call communities on. Is there a convenient way to convert either of these objects to a `networkit.graph.Graph`?

Thanks!

**Comment by clstaudt** [comment]
I don't think there is - but there should be. Would be good to realize ideas for better `numpy`/`scipy` integration.

Of course you can always iterate over the matrix and build the graph that way.

**Comment by hmeyerhenke** [comment]
If it's a sparse matrix, 60K is not so large. An incremental build would be reasonably fast. We look into this feature request as well for a simpler solution, but this may take some time.

**Comment by Jessime** [comment]
Just to clarify, I'm looking at dealing with a 60K-by-60K matrix. I don't know the exact density, but if we assume ~10%, I have around  423M elements. 

I used some simulated data (`sim`) and did:

```python
g = networkit.graph.Graph(65000)
zipped = zip(product(range(6500), range(65000)), np.nditer(sim))
for (i, j), w in tqdm_notebook(zipped, total=sim.size):
    g.addEdge(i, j, w)
```

Iterating over 423M edges took about 6.5 minutes. So, not terrible, really.

**Comment by hmeyerhenke** [comment]
That's a pretty large fill-in indeed. Great if you can live with this solution.

Anyone out there with a feeling how much faster a more direct conversion would be?

**Comment by saibalmars** [comment]
Hi all,
Here is my script to convert `scipy.sparse.csr_matrix` to Networkit graph that also suitable for weighted or directed graphs. Hope this helps.

```python
def sparse2nk(A):
    """
    Convert scipy.sparse csr_matrix into Networkit graph
    :param A: `scipy.sparse.csr_matrix` matrix
    :return: Networkit graph
    """
    # check if graph is weighted
    if A.data.max() == A.data.min() == 1:
        is_weighted = False
    else:
        is_weighted = True

    # check if graph is directed
    if (abs(A.T - A) > 1e-10).nnz == 0:  # if A is symmetry
        is_directed = False
    else:
        is_directed = True

    num_nodes = A.shape[0]

    nkG = nk.graph.Graph(num_nodes, weighted=is_weighted, directed=is_directed)

    for x, y in zip(A.nonzero()[0], A.nonzero()[1]):
        if x < y and not is_directed:
            pass
        else:
            nkG.addEdge(x, y, A[x, y]) if is_weighted else nkG.addEdge(x, y)

    return nkG
```

**Comment by angriman** [comment]
Thank you. Similarly to NetworKit `nxadapter.py`, I think that it would be a good idea to create a new `networkit/scipyadapter.py` file with this code. In addition to `sparse2nk`, we should also provide the "inverse" `nk2sparse` function.

**Comment by clstaudt** [comment]
To optimize performance, consider
- making the check for directedness optional
- replacing  `zip` by `numpy` array operations, e.g. 
         https://docs.scipy.org/doc/numpy/reference/generated/numpy.dstack.html

And consider whether one function could also easily handle other types of scipy sparse matrix as well as numpy dense arrays .

**Comment by avdgrinten** [comment]
Actually, this is probably easier than I thought. Cython's numpy integration should help to make this code very efficient. This is something @cndolo can look into in the future (but in the short term, other refactoring has a higher priority).
